### PR TITLE
feat: Implement the Buf to avoid extra memory allocation

### PR DESCRIPTION
**Pull request by LYZJU2019**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#4065 

## What's changed and what's your intention?
implement Buffer to avoid extra memory allocation in file `src/mito2/src/wal/entry_reader.rs`.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

**Comment by WenyXu** [comment]
The extra memory allocation happened in the `concat` method. We should move the `parts`(`Vec<Vec<u8>>`) in `MultiplePartEntry` into the `Reader`.
https://github.com/GreptimeTeam/greptimedb/blob/4e5dd1ebb04d910bdb88611a8f824529701606b8/src/store-api/src/logstore/entry.rs#L142

**Comment by WenyXu** [comment]
The ` Toml Check` is failed, run `make fmt-toml` to fix it.

File: src/mito2/Cargo.toml [hunk 1]
```diff
@@ -46,6 +46,7 @@ log-store = { workspace = true, optional = true }
 memcomparable = "0.2"
 moka = { workspace = true, features = ["sync", "future"] }
 object-store.workspace = true
+opendal = { version = "*" }
```
[the remark this answers is not in the document]
Comment (LYZJU2019) [reply src/mito2/Cargo.toml:49]: I think we should re-export `Buffer` via `store-api` crate instead of `object-store`?

**Comment by LYZJU2019** [comment]
Are we supposed to change the return value type of into_bytes() method? I cannot think of a way to avoid extra memory allocation when we have to flat a 2d array into an 1d array.

**Comment by WenyXu** [comment]
> Are we supposed to change the return value type of into_bytes() method? 

Yes

File: src/mito2/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (LYZJU2019) [reply src/mito2/Cargo.toml:49]: should it be `store-api` crate instead of `object-store` crate?

[the remark this answers is not in the document]
Comment (WenyXu) [reply src/mito2/Cargo.toml:49]: How about letting the `into_bytes` to return an enum and importing the `object-store` in the `mito` crate?

[the remark this answers is not in the document]
Comment (killme2008) [reply src/mito2/Cargo.toml:49]: @LYZJU2019 Hi, thank you for your contribution. Could you refactor it as @WenyXu said? Let's merge this PR after the refactoring.

**Comment by WenyXu** [comment]
completed in #4585

**Pull request closed without merging**